feat: auto upgrade ecs task role

edgar-castle opened the pull request:
### Feature/bug description:

Adds support for auto-upgrading ECS tasks to our newly defined IAM roles.

I'll hold off on merging anything after approvals to ensure engineering wide comms go out before releasing the change.

### This is how I decided to implement/fix it:

* [`client/client.go`](diffhunk://#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09R12-R56): Introduced `ecsInterface` and `iamInterface` to abstract ECS and IAM client functionalities, added support for `dryRun` and `taskRoleArn` parameters, and updated methods to use the new interfaces. [[1]](diffhunk://#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09R12-R56) [[2]](diffhunk://#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09R69-R79) [[3]](diffhunk://#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09L56-R95) [[4]](diffhunk://#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09L75-R120) [[5]](diffhunk://#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09L117-R162) [[6]](diffhunk://#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09L126-R171)
* [`client/task_role_upgrade.go`](diffhunk://#diff-0205746dc6acec21b24a49b03e68e5b8e8b1183c7d6be1b0fb503c5b28de7ac0R1-R97): Added a new file to handle task role upgrades, including functions to extract service information and verify IAM roles.
* [`client/task_role_upgrade_test.go`](diffhunk://#diff-1463292df69db0861f69d7dff3fd31a395cdb92b2605220c5996b4515c3f84bdR1-R290): Added comprehensive tests for the new task role upgrade functionality.

Dependency updates and configuration changes:

* [`go.mod`](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L3-R16): Updated Go version to 1.22 and added new dependencies for testing and YAML processing.
* [`.codacy.yml`](diffhunk://#diff-6e27af35533eda2e3e82d6d7f8654c26824396797716d37ac7df096a1b090503L1-L8): Removed exclusion paths for test files and markdown files.

New configuration options:

* [`main.go`](diffhunk://#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R19-R23): Added flags for `task-role-arn` and `dry-run` to provide more control over task definitions and deployment processes. [[1]](diffhunk://#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R19-R23) [[2]](diffhunk://#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261L44-R46) [[3]](diffhunk://#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R60-R62) [[4]](diffhunk://#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261L76-R81)

### What can this break:
Our ECS deploys

### How has this been tested:
Added in unit tests for the new functionality added in this PR.

Details on integration testing can be found [here](https://www.notion.so/gametime/ecs-deploy-role-upgrader-1cb898c3e3ea801693d5fd65f8df2604).

edgar-castle committed:
feat: auto upgrade ecs task role

Signed-off-by: Edgar Castillo <edgar.castillo@gametime.co>

edgar-castle committed:
feat: add regex validation on client.taskRoleArn

Signed-off-by: Edgar Castillo <edgar.castillo@gametime.co>

waitstime commented on client/task_role_upgrade.go:
```diff
@@ -0,0 +1,97 @@
+func (c *Client) getTaskRole(currentRoleArn, service *string) (*string, error) {
+	// If a valid task role ARN is set in the client, return that
+	if c.taskRoleArn != "" {
+		if roleRegex.MatchString(c.taskRoleArn) {
+			c.logger.Printf("[info] Using supplied task role ARN: %s", c.taskRoleArn)
+			return &c.taskRoleArn, nil
+		}
```
Should this be checking the `taskRoleRegex` instead since that's what we are migrating to? Or maybe check if it matches either `roleRegex` or `taskRoleRegex` since this is the task role Arn that someone is passing in so we should assume they are correctly selecting their role. Or perhaps we should just take the role that's specified at face value and not perform checks.

edgar-castle replied:
I can update that comment to be clearer. That check will return the set task role for this condition [here](https://www.notion.so/gametime/ecs-deploy-role-upgrader-1cb898c3e3ea801693d5fd65f8df2604?pvs=4#1cb898c3e3ea80d2aa2dcd52c6822006).

waitstime replied:
I'm cool with this then. Maybe you don't need a regex check at all and can just set it to what was passed. But the regex will probably work for most everything that's not the new format.

edgar-castle replied:
I actually had it like that at first but figured I'd add some validation for it.

In case someone mistypes the role arn, it can help them debug the issue as it'll log the error.

davekonopka reviewed (approved):
The code and test coverage all look good to me. Functionality makes sense. Only questions I have are operational -- want to ensure this will work with the Go version bump on Travis and that everywhere this is used the AWS auth context has IAM read access.

edgar-castle committed:
refactor: use std lib for testing to avoid introducing toolchain directive

Signed-off-by: Edgar Castillo <edgar.castillo@gametime.co>

edgar-castle committed:
Merge pull request #119 from gametimesf/use-std-lib-for-testing

refactor: use std lib for testing to avoid introducing toolchain dir

edgar-castle commented:
> The code and test coverage all look good to me. Functionality makes sense. Only questions I have are operational -- want to ensure this will work with the Go version bump on Travis and that everywhere this is used the AWS auth context has IAM read access.

Thanks Dave, very good points. I verified the auth side and removed the need for the go version bump by using the std lib for testing.

Good news is, if there's any error in the functionality I'm introducing, it'll return the current task role arn, resulting in a no-op on the role update side. So in the event we're missing the needed permission somewhere, it'll re-use the current task role and log the corresponding error.

I tested it out and added in an example of that [here](https://www.notion.so/gametime/ecs-deploy-role-upgrader-1cb898c3e3ea801693d5fd65f8df2604?pvs=4#1d1898c3e3ea8089b2a7f4001e5d23bb).

waitstime commented on client/task_role_upgrade.go:
```diff
@@ -0,0 +1,97 @@
+		return "", "", fmt.Errorf("service name does not match expected format")
+	}
+
+	service = strings.TrimSuffix(service, "-service")
+	env := strings.Split(service, "-")[0]
+	name := strings.TrimPrefix(service, env+"-")
```
Or is the plan to manually update the one's that don't fit the mold. It's probably less than 10

edgar-castle replied:
Yeah exactly, this won't cover 100% of our services but should cover like 90%+. I plan on manually tackling those remaining edge cases.

If the service name doesn't conform already, the function will return early at line 64 during the regex matching.

waitstime reviewed (approved):
Should release as 2.0 even though we use `latest` every where.
